server/src/engine.rs

```diff
@@ -1,33 +1,102 @@
+    MultipleServiceTypes,
 
-    match ServiceType::from_i32(query.service_type) {
+    /// Request contains negative weight.
+    #[snafu(display("Request contains negative weight"))]
+    NegativeWeight,
+
+    /// Request contains negative distance.
+    #[snafu(display("Request contains negative distance"))]
+    NegativeDistance,
+
+    /// No pricing requests were provided.
+    #[snafu(display("No pricing requests were provided"))]
+    NoRequests,
+
+    /// Unknown service type error.
+    #[snafu(display("Unknown service type; cannot parse service type"))]
+    UnknownServiceType,
+}
+
+/// Get pricing given a [`PricingRequests`], which contains an array of
+/// [`PricingRequest`]s.
+///
+/// The array of [`PricingRequest`]s is used to allow pricing for
+/// multiple legs of a trip to be calculated.
+///
+/// # Notes
+/// The `service_type` field of all [`PricingRequest`]s in the array
+/// must be the same. It means the pricing service currently does not
+/// allow for a trip that involves different types of services.
+///
+/// This may change in the future as per business requirements.
+///
+/// # Returns
+/// If no errors occur, returns a vector of prices in dollars,
+/// corresponding to the order of the input requests.
+pub fn get_pricing(query: PricingRequests) -> Result<Vec<f32>, PricingError> {
+    let requests = query.requests;
+    info!("Getting pricing for {:?} requests", requests.len());
+    check_pricing_requests(&requests)?;
+    let mut prices = Vec::new();
+    match ServiceType::from_i32(requests[0].service_type) {
         Some(ServiceType::Cargo) => {
             debug!("Cargo pricing");
-            get_cargo_pricing(query)
+            for request in requests {
+                prices.push(get_cargo_pricing(request));
+            }
+            Ok(prices)
         }
         Some(ServiceType::Rideshare) => {
             debug!("Rideshare pricing");
-            get_rideshare_pricing(query)
+            for request in requests {
+                prices.push(get_rideshare_pricing(request));
+            }
+            Ok(prices)
         }
         Some(ServiceType::Charter) => {
             debug!("Charter pricing");
-            get_charter_pricing(query)
+            for request in requests {
+                prices.push(get_charter_pricing(request));
+            }
+            Ok(prices)
         }
         None => {
-            debug!("Error parsing service type; returning 0.0");
-            0.0
+            debug!("Error parsing service type");
+            Err(PricingError::UnknownServiceType)
         }
     }
 }
 
+/// Check that all pricing requests have the same service type.
+fn check_pricing_requests(requests: &[PricingRequest]) -> Result<(), PricingError> {
+    if requests.is_empty() {
+        return Err(PricingError::NoRequests);
+    }
+    let service_type = requests[0].service_type;
+    for request in requests {
+        if request.service_type != service_type {
```

Comment:
This is fine for now -

Do you think we'll ever have flights that are carrying cargo _and_ passengers? If there's passenger seating and a small cargo hold, for instance, for people and objects going to the same vertiport

Reply:
I suppose this calculation is for a specific customer order, and their order shouldn't turn from a cargo mission to a passenger mission between legs. Ignore my comment :)

Reply:
yea this is purely to prevent any programming errors when making rpc calls to pricing.